### ISSUE TITLE
FreeBSD Platform: Xemu Port 

### ISSUE THREAD

**Issue by hernandp** [lgblgblgb/xemu#110]
FreeBSD is the most popular BSD UNIX system, running in several architectures, and a great development platform. Its port system makes auto compilation of "foreign" projects easily.  Just to expand the systems where Xemu runs on, I would like to do a nice port to FreeBSD. 
Of course, I would remain as a mantainer for it.

**Comment by lgblgblgb** [comment]
It cannot hurt :) I still set the "low priority" label, if you don't mind, but indeed, it's a fun idea.

**Comment by lgblgblgb** [comment]
OK, I've tried. It compiled without any modification on FreeBSD, haven't tried yet to start it :) However some notes:

1. of course GNU make (`gmake`) is needed, it was expected
2. /bin/bash must exist at that PATH, it was a problem for me, since from ports it installs to /usr/local/bin  I simply symlinked to /bin/bash and it worked.

**Comment by lgblgblgb** [comment]
Unfortunately I cannot try to RUN it, since VirtualBox has some problem with FreeBSD that no keyboard or mouse works in X :-O It seems to be a common issue (according to Google) but whatever I tried as solution, didn't work. So I can't use X with FreeBSD to test it, just the compilation itself. Though as an interesting fact, I could start Xemu with having a timed startup script from .xinitrc :) Then it segfaults somewhere in the SDL init :-O More work is needed to test this with some sane FreeBSD installation unlike mine one, tried just for fun ...

**Comment by hernandp** [comment]
> Unfortunately I cannot try to RUN it, since VirtualBox has some problem with FreeBSD that no keyboard or mouse works in X :-O It seems to be a common issue (according to Google) but whatever I tried as solution, didn't work. So I can't use X with FreeBSD to test it, just the compilation itself. Though as an interesting fact, I could start Xemu with having a timed startup script from .xinitrc :) Then it segfaults somewhere in the SDL init :-O More work is needed to test this with some sane FreeBSD installation unlike mine one, tried just for fun ...

Don't worry, my FreeBSD has X installed, I will test it ASAP.

**Comment by lgblgblgb** [comment]
_Btw, as an off-topic here, if you have any idea how to make FreeBSD work as a guest within VirtualBox, I'm open to hear :) I've tried everything, installed virtualbox tools in FreeBSD, tried moused, hald, dbus, everything in every combination, with xorg configs in /etc/X11, etc, always the same, the mouse cursor moves, but clicks does not work and no keyboard input either. The keyboard works on the text consoles for sure ..._

**Comment by hernandp** [comment]
We can say that FreeBSD status (Release 12.0) is that xemu can be built by installing the packages or ports of  `gmake` , `bash`, `sdl2` and using `CC=clang`  to switch to the native `clang` compiler.

But I could not solve SDL segfaulting at startup under a VMWare machine, even with vmware driver, and acceleration enabled in Xorg.

**Comment by hernandp** [comment]
> _Btw, as an off-topic here, if you have any idea how to make FreeBSD work as a guest within VirtualBox, I'm open to hear :) I've tried everything, installed virtualbox tools in FreeBSD, tried moused, hald, dbus, everything in every combination, with xorg configs in /etc/X11, etc, always the same, the mouse cursor moves, but clicks does not work and no keyboard input either. The keyboard works on the text consoles for sure ..._

IT didnt work until I put `vmmouse` as driver and a custom X.org (autoconfig did not detect anything other than standard mouse).

Did you try to override autodetection with

```
Section "ServerLayout"
...
Option       "AutoAddDevices" "Off"
...
EndSection
```

**Comment by lgblgblgb** [comment]
Btw, does using "cc" as the compiler also works? By tradition the C compiler in a UNIX system can be launched with just cc, usually it's a symlink or something to the actual C compiler in a modern UNIX system like to gcc or clang. I guess I would need to change the default "gcc" to simply cc. So on systems having gcc either clang would work then without too much issue. Also it should be checked that native Windows build using MSYS also works this way, ie, just saying "cc" is enough then. Also the same on OSX, that saying "cc" works as well. If that also works, I think ALL native compilation should use cc. Surely cross-compilation is a bit different story!

**Comment by hernandp** [comment]
> Btw, does using "cc" as the compiler also works? By tradition the C compiler in a UNIX system can be launched with just cc, usually it's a symlink or something to the actual C compiler in a modern UNIX system like to gcc or clang. I guess I would need to change the default "gcc" to simply cc. So on systems having gcc either clang would work then without too much issue. Also it should be checked that native Windows build using MSYS also works this way, ie, just saying "cc" is enough then. Also the same on OSX, that saying "cc" works as well. If that also works, I think ALL native compilation should use cc. Surely cross-compilation is a bit different story!

BTW, I checked that FreeBSD and MSYS2 in Windows both include `cc` in base installation. Maybe we can use `cc` as the standard in the build tools.

**Comment by lgblgblgb** [comment]
Is it true for MacOS as well? I mean, cc works there also, and does it start the very same thing gcc would? (which is odd in case of Mac, since for real it's an apple-modified clang, but for some reason they name it as "gcc" as well ... but anyway).

**Comment by hernandp** [comment]
> Is it true for MacOS as well? I mean, cc works there also, and does it start the very same thing gcc would? (which is odd in case of Mac, since for real it's an apple-modified clang, but for some reason they name it as "gcc" as well ... but anyway).

Yes, macOS also links `cc` to `clang`.

**Comment by lgblgblgb** [comment]
OK, what I should do with this ticket? Can I close it, or any kind of value to have it around open?

**Comment by hernandp** [comment]
I think we can close it. We know that compiling under FBSD is effortless with the current makefiles and build system, so no porting is necessary.

**Comment by lgblgblgb** [comment]
OK, so closing! And thanks.

**Comment by lgblgblgb** [comment]
It seems Xemu segfaults sometimes (always?) on FreeBSD. The issue is VERY strange. It seems there are multiple calls for hid_init(), and the first is by NULL pointer as one of its parameters. It's very much a mystery for me, since that is called only ONCE and never with NULL, so in theory it's impossible. Anyway, I try to handle this by checking the problematic pointer ...
